### PR TITLE
Remove funcref as a subtype of anyref

### PR DESCRIPTION
**Pull request by eqrion**
This is the equivalent of the syntax changes in WebAssembly/reference-types/pull/87. I have a corresponding patch in SpiderMonkey that tests this out with the semantics changes as well.

It may be good to hold off on merging this until the upstream PR is closer to merging, to track any changes. I'm posting this now to just get ready for when that is.

### CODE REVIEW

File: crates/wast/src/ast/types.rs [hunk 1]
```diff
@@ -49,52 +42,27 @@ impl<'a> Parse<'a> for ValType<'a> {
...
         } else if l.peek::<kw::funcref>() {
             parser.parse::<kw::funcref>()?;
-            Ok(ValType::Funcref)
+            Ok(ValType::Ref(RefType::Func))
         } else if l.peek::<kw::anyfunc>() {
             parser.parse::<kw::anyfunc>()?;
-            Ok(ValType::Funcref)
+            Ok(ValType::Ref(RefType::Func))
+        } else if l.peek::<kw::r#externref>() {
+            parser.parse::<kw::r#externref>()?;
```
Comment (alexcrichton) [inline crates/wast/src/ast/types.rs:50]: This and above I think can drop the `r#` prefix

File: crates/wast/src/ast/types.rs [hunk 2]
```diff
@@ -104,13 +72,68 @@ impl<'a> Parse<'a> for ValType<'a> {
...
+            Err(l.error())
+        }
+    }
+}
+
+/// The reference value types for a wasm module.
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+pub enum RefType<'a> {
+    Func,
```
Comment (alexcrichton) [inline crates/wast/src/ast/types.rs:92]: Mind adding a few comments to the variants here? The first two I think are reference-types proposal types, the next is exception handling, and everything else is the gc proposal, right?

File: crates/wast/src/ast/types.rs [hunk 3]
```diff
@@ -104,13 +72,68 @@ impl<'a> Parse<'a> for ValType<'a> {
...
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let mut l = parser.lookahead1();
+        if l.peek::<kw::func>() {
+            parser.parse::<kw::func>()?;
+            Ok(RefType::Func)
+        } else if l.peek::<kw::r#extern>() {
+            parser.parse::<kw::r#extern>()?;
+            Ok(RefType::Extern)
+        } else if l.peek::<kw::any>() {
+            parser.parse::<kw::any>()?;
```
Comment (alexcrichton) [inline crates/wast/src/ast/types.rs:121]: Are there any old tests using `(ref any)` after everything is renamed?

File: crates/wast/src/ast/types.rs (hunk 2 above)
Comment (eqrion) [reply crates/wast/src/ast/types.rs:94]: Yes, that's the case. I've added some comments explaining the types.

File: crates/wast/src/ast/types.rs (hunk 3 above)
Comment (eqrion) [reply crates/wast/src/ast/types.rs]: No, there shouldn't really be any. Might be some in wabt, but they'll update at some point. I've removed the alias for this form.

File: ' [hunk 4]
```diff
@@ -0,0 +1,31 @@
+r cd9f003 Drop nullref type
```
Comment (alexcrichton) [inline ':1]: Stray file added in?

Comment (eqrion) [reply ']: Oh, bizarre. I've no idea how my rebase commands got committed. Nice catch!

**Comment by eqrion** [comment]
Yes, let's merge this and get a new version when you get a chance!

**Pull request merged**